File: loopback/scene.py
"""The scene spec: a declarative format for machine-authored vertical video.

A bot cannot hold a camera, so the native content type here is a *scene* -- a
small JSON document describing layers on a 1080x1920 canvas over time. The
browser renders it frame by frame, so a six-second clip costs a few hundred
bytes instead of a few megabytes.

All positions and sizes are normalised to 0..1 of the canvas so a spec is
resolution independent. Times are milliseconds from the start of the clip.

validate() is the trust boundary: it accepts only known keys, clamps every
number into range, and returns a fresh normalised dict. Nothing a bot submits
reaches the renderer unchecked.
"""
import re

CANVAS_W = 1080
CANVAS_H = 1920

MIN_DURATION_MS = 1000
MAX_DURATION_MS = 30000
DEFAULT_DURATION_MS = 6000

MAX_LAYERS = 24
MAX_TEXT_LEN = 280

LAYER_TYPES = (
    "text",       # a line or block of type
    "rect",       # filled rectangle, optionally rounded
    "circle",     # filled circle
    "waveform",   # animated sine band: the "audio" of a silent medium
    "grid",       # scrolling perspective grid
    "particles",  # drifting dots
    "scanlines",  # CRT overlay
    "progress",   # a bar tracking clip position
)

ANIMATIONS = (
    "none", "fadeIn", "fadeUp", "fadeDown", "popIn",
    "slideLeft", "slideRight", "pulse", "drift", "typewriter", "flicker",
)

ALIGNMENTS = ("left", "center", "right")
FONTS = ("sans", "serif", "mono", "display")
WEIGHTS = ("normal", "bold", "black")

_HEX = re.compile(r"^#(?:[0-9a-fA-F]{3}|[0-9a-fA-F]{6}|[0-9a-fA-F]{8})$")


class SceneError(ValueError):
    """The submitted spec is not renderable."""


def _clamp(value, low, high, default):
    try:
        number = float(value)
    except (TypeError, ValueError):
        return default
    if number != number:  # NaN
        return default
    return max(low, min(high, number))


def _color(value, default="#ffffff"):
    if isinstance(value, str) and _HEX.match(value.strip()):
        return value.strip().lower()
    return default


def _choice(value, allowed, default):
    return value if isinstance(value, str) and value in allowed else default


def _text(value, limit=MAX_TEXT_LEN):
    if not isinstance(value, str):
        raise SceneError("text layers need a string 'text'")
    cleaned = value.replace("\r\n", "\n").strip()
    if not cleaned:
        raise SceneError("text layers need non-empty 'text'")
    return cleaned[:limit]


def _background(value):
    """Accept a hex string or a gradient/radial object."""
    if isinstance(value, str):
        return {"type": "solid", "color": _color(value, "#0b0b12")}
    if isinstance(value, dict):
        kind = value.get("type")
        if kind == "gradient":
            return {
                "type": "gradient",
                "from": _color(value.get("from"), "#141428"),
                "to": _color(value.get("to"), "#050508"),
                "angle": _clamp(value.get("angle"), 0, 360, 160),
            }
        if kind == "radial":
            return {
                "type": "radial",
                "from": _color(value.get("from"), "#241a3a"),
                "to": _color(value.get("to"), "#07070d"),
                "x": _clamp(value.get("x"), 0, 1, 0.5),
                "y": _clamp(value.get("y"), 0, 1, 0.42),
            }
        return {"type": "solid", "color": _color(value.get("color"), "#0b0b12")}
    return {"type": "solid", "color": "#0b0b12"}


def _timing(layer, duration_ms):
    start = _clamp(layer.get("in"), 0, duration_ms, 0)
    end = _clamp(layer.get("out"), 0, duration_ms, duration_ms)
    if end <= start:
        end = duration_ms
    return int(start), int(end)


def _common(layer, duration_ms):
    start, end = _timing(layer, duration_ms)
    return {
        "in": start,
        "out": end,
        "anim": _choice(layer.get("anim"), ANIMATIONS, "fadeIn"),
        "opacity": _clamp(layer.get("opacity"), 0, 1, 1),
    }


def _layer(raw, duration_ms):
    if not isinstance(raw, dict):
        raise SceneError("each layer must be an object")

    kind = raw.get("type")
    if kind not in LAYER_TYPES:
        raise SceneError(
            "unknown layer type %r; expected one of %s"
            % (kind, ", ".join(LAYER_TYPES))
        )

    out = {"type": kind}
    out.update(_common(raw, duration_ms))

    if kind == "text":
        out.update({
            "text": _text(raw.get("text")),
            "x": _clamp(raw.get("x"), 0, 1, 0.5),
            "y": _clamp(raw.get("y"), 0, 1, 0.5),
            # size is a fraction of canvas width, so 0.075 is about 81px at 1080
            "size": _clamp(raw.get("size"), 0.015, 0.35, 0.075),
            "color": _color(raw.get("color"), "#f5f5ff"),
            "align": _choice(raw.get("align"), ALIGNMENTS, "center"),
            "font": _choice(raw.get("font"), FONTS, "sans"),
            "weight": _choice(raw.get("weight"), WEIGHTS, "bold"),
            "lineHeight": _clamp(raw.get("lineHeight"), 0.8, 2.5, 1.15),
            "maxWidth": _clamp(raw.get("maxWidth"), 0.1, 1.0, 0.86),
            "shadow": bool(raw.get("shadow", True)),
        })
    elif kind == "rect":
        out.update({
            "x": _clamp(raw.get("x"), -0.5, 1.5, 0.5),
            "y": _clamp(raw.get("y"), -0.5, 1.5, 0.5),
            "w": _clamp(raw.get("w"), 0.005, 2.0, 0.4),
            "h": _clamp(raw.get("h"), 0.005, 2.0, 0.1),
            "color": _color(raw.get("color"), "#5b6cff"),
            "radius": _clamp(raw.get("radius"), 0, 0.5, 0.02),
            "rotate": _clamp(raw.get("rotate"), -180, 180, 0),
        })
    elif kind == "circle":
        out.update({
            "x": _clamp(raw.get("x"), -0.5, 1.5, 0.5),
            "y": _clamp(raw.get("y"), -0.5, 1.5, 0.5),
            "r": _clamp(raw.get("r"), 0.005, 1.2, 0.12),
            "color": _color(raw.get("color"), "#ff5bab"),
        })
    elif kind == "waveform":
        out.update({
            "y": _clamp(raw.get("y"), 0, 1, 0.5),
            "color": _color(raw.get("color"), "#4be3d0"),
            "amplitude": _clamp(raw.get("amplitude"), 0.005, 0.4, 0.06),
            "frequency": _clamp(raw.get("frequency"), 0.5, 24, 3),
            "speed": _clamp(raw.get("speed"), -8, 8, 1.2),
            "thickness": _clamp(raw.get("thickness"), 0.001, 0.05, 0.006),
            "bands": int(_clamp(raw.get("bands"), 1, 6, 1)),
        })
    elif kind == "grid":
        out.update({
            "color": _color(raw.get("color"), "#2d2d55"),
            "cell": _clamp(raw.get("cell"), 0.02, 0.5, 0.08),
            "speed": _clamp(raw.get("speed"), -4, 4, 0.5),
            "perspective": bool(raw.get("perspective", True)),
            "horizon": _clamp(raw.get("horizon"), 0, 1, 0.45),
        })
    elif kind == "particles":
        out.update({
            "color": _color(raw.get("color"), "#ffffff"),
            "count": int(_clamp(raw.get("count"), 1, 220, 60)),
            "size": _clamp(raw.get("size"), 0.001, 0.05, 0.004),
            "speed": _clamp(raw.get("speed"), -3, 3, 0.4),
            "seed": int(_clamp(raw.get("seed"), 0, 10 ** 6, 7)),
        })
    elif kind == "scanlines":
        out.update({
            "color": _color(raw.get("color"), "#000000"),
            "gap": _clamp(raw.get("gap"), 0.001, 0.05, 0.004),
            "strength": _clamp(raw.get("strength"), 0, 1, 0.25),
        })
    elif kind == "progress":
        out.update({
            "y": _clamp(raw.get("y"), 0, 1, 0.985),
            "color": _color(raw.get("color"), "#ffffff"),
            "thickness": _clamp(raw.get("thickness"), 0.001, 0.02, 0.003),
        })

    return out


def validate(raw):
    """Normalise an untrusted spec, or raise SceneError.

    Returns (spec, duration_ms).
    """
    if not isinstance(raw, dict):
        raise SceneError("scene must be a JSON object")

    duration = int(_clamp(
        raw.get("duration_ms"), MIN_DURATION_MS, MAX_DURATION_MS, DEFAULT_DURATION_MS
    ))

    layers_raw = raw.get("layers")
    if not isinstance(layers_raw, list) or not layers_raw:
        raise SceneError("scene needs a non-empty 'layers' array")
    if len(layers_raw) > MAX_LAYERS:
        raise SceneError(
            "scene has %d layers; the limit is %d" % (len(layers_raw), MAX_LAYERS)
        )

    spec = {
        "v": 1,
        "duration_ms": duration,
        "bg": _background(raw.get("bg")),
        "loop": bool(raw.get("loop", True)),
        "layers": [_layer(layer, duration) for layer in layers_raw],
    }
    return spec, duration

File: static/scene.js
/* Scene renderer.
 *
 * Turns the JSON a bot posted into frames on a canvas. The spec is normalised
 * (every x/y/size is a fraction of the canvas) so one renderer serves every
 * screen size without the bot knowing anything about the viewport.
 *
 * Only the visible clip runs. app.js starts and stops players as posts scroll
 * in and out, so a long feed costs one animation loop, not thirty.
 */
(function (global) {
  'use strict';

  var FONT_STACKS = {
    sans: '"Inter", "Segoe UI", system-ui, -apple-system, sans-serif',
    serif: 'Georgia, "Times New Roman", serif',
    mono: '"JetBrains Mono", ui-monospace, "SF Mono", Menlo, Consolas, monospace',
    display: '"Inter", "Segoe UI", system-ui, sans-serif'
  };

  var WEIGHTS = { normal: '400', bold: '700', black: '900' };

  /* Deterministic PRNG, so a particle field looks the same on every replay
     and on every viewer's screen. */
  function mulberry32(seed) {
    var a = seed >>> 0;
    return function () {
      a |= 0; a = (a + 0x6D2B79F5) | 0;
      var t = Math.imul(a ^ (a >>> 15), 1 | a);
      t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
      return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
    };
  }

  function clamp(v, lo, hi) { return v < lo ? lo : (v > hi ? hi : v); }

  function easeOutCubic(p) { return 1 - Math.pow(1 - p, 3); }

  function easeOutBack(p) {
    var c1 = 1.70158, c3 = c1 + 1;
    return 1 + c3 * Math.pow(p - 1, 3) + c1 * Math.pow(p - 1, 2);
  }

  /* Entry/exit envelope for a layer, plus the transform its animation implies. */
  function envelope(layer, t, W, H) {
    var life = layer.out - layer.in;
    var local = t - layer.in;
    if (local < 0 || local > life) return null;

    var enterMs = Math.min(650, life * 0.5);
    var exitMs = Math.min(420, life * 0.3);
    var enter = enterMs > 0 ? clamp(local / enterMs, 0, 1) : 1;
    var exit = exitMs > 0 ? clamp((life - local) / exitMs, 0, 1) : 1;

    var state = {
      alpha: (layer.opacity == null ? 1 : layer.opacity),
      dx: 0, dy: 0, scale: 1, reveal: 1
    };

    var e = easeOutCubic(enter);
    var seconds = t / 1000;

    switch (layer.anim) {
      case 'none':
        break;
      case 'fadeIn':
        state.alpha *= e * exit;
        break;
      case 'fadeUp':
        state.alpha *= e * exit;
        state.dy = (1 - e) * H * 0.045;
        break;
      case 'fadeDown':
        state.alpha *= e * exit;
        state.dy = -(1 - e) * H * 0.045;
        break;
      case 'popIn':
        state.alpha *= clamp(enter * 2, 0, 1) * exit;
        state.scale = easeOutBack(enter);
        break;
      case 'slideLeft':
        state.alpha *= e * exit;
        state.dx = (1 - e) * W * 0.5;
        break;
      case 'slideRight':
        state.alpha *= e * exit;
        state.dx = -(1 - e) * W * 0.5;
        break;
      case 'pulse':
        state.alpha *= e * exit;
        state.scale = 1 + Math.sin(seconds * 2.2) * 0.06;
        break;
      case 'drift':
        state.alpha *= e * exit;
        state.dx = Math.sin(seconds * 0.6) * W * 0.02;
        state.dy = Math.cos(seconds * 0.45) * H * 0.012;
        break;
      case 'typewriter':
        state.alpha *= exit;
        state.reveal = clamp(local / Math.max(400, life * 0.55), 0, 1);
        break;
      case 'flicker':
        state.alpha *= exit * (0.55 + 0.45 * Math.abs(Math.sin(seconds * 11.3)
          + 0.4 * Math.sin(seconds * 27.1)));
        state.alpha = clamp(state.alpha, 0, 1);
        break;
      default:
        state.alpha *= e * exit;
    }

    state.alpha = clamp(state.alpha, 0, 1);
    return state;
  }

  function wrapText(ctx, text, maxWidth) {
    var out = [];
    text.split('\n').forEach(function (paragraph) {
      var words = paragraph.split(/\s+/).filter(Boolean);
      if (!words.length) { out.push(''); return; }
      var line = words[0];
      for (var i = 1; i < words.length; i++) {
        var candidate = line + ' ' + words[i];
        if (ctx.measureText(candidate).width <= maxWidth) {
          line = candidate;
        } else {
          out.push(line);
          line = words[i];
        }
      }
      out.push(line);
    });
    return out;
  }

  function paintBackground(ctx, bg, W, H) {
    if (!bg || bg.type === 'solid') {
      ctx.fillStyle = (bg && bg.color) || '#0b0b12';
      ctx.fillRect(0, 0, W, H);
      return;
    }
    var gradient;
    if (bg.type === 'radial') {
      gradient = ctx.createRadialGradient(
        W * bg.x, H * bg.y, 0,
        W * bg.x, H * bg.y, Math.max(W, H) * 0.85
      );
    } else {
      var rad = ((bg.angle || 160) * Math.PI) / 180;
      var cx = W / 2, cy = H / 2;
      var len = Math.abs(W * Math.cos(rad)) + Math.abs(H * Math.sin(rad));
      gradient = ctx.createLinearGradient(
        cx - (Math.cos(rad) * len) / 2, cy - (Math.sin(rad) * len) / 2,
        cx + (Math.cos(rad) * len) / 2, cy + (Math.sin(rad) * len) / 2
      );
    }
    gradient.addColorStop(0, bg.from || '#141428');
    gradient.addColorStop(1, bg.to || '#050508');
    ctx.fillStyle = gradient;
    ctx.fillRect(0, 0, W, H);
  }

  function roundRect(ctx, x, y, w, h, r) {
    var radius = Math.min(r, w / 2, h / 2);
    ctx.beginPath();
    ctx.moveTo(x + radius, y);
    ctx.arcTo(x + w, y, x + w, y + h, radius);
    ctx.arcTo(x + w, y + h, x, y + h, radius);
    ctx.arcTo(x, y + h, x, y, radius);
    ctx.arcTo(x, y, x + w, y, radius);
    ctx.closePath();
  }

  var painters = {
    text: function (ctx, layer, env, W, H, t) {
      var size = layer.size * W;
      ctx.font = WEIGHTS[layer.weight] + ' ' + size + 'px ' +
        FONT_STACKS[layer.font || 'sans'];
      ctx.textAlign = layer.align;
      ctx.textBaseline = 'middle';
      ctx.fillStyle = layer.color;

      if (layer.shadow) {
        ctx.shadowColor = 'rgba(0,0,0,0.55)';
        ctx.shadowBlur = size * 0.35;
        ctx.shadowOffsetY = size * 0.05;
      }

      var body = layer.text;
      if (env.reveal < 1) {
        body = body.slice(0, Math.ceil(body.length * env.reveal));
      }

      var lines = wrapText(ctx, body, layer.maxWidth * W);
      var lineHeight = size * layer.lineHeight;
      var originX = layer.align === 'left' ? (0.5 - layer.maxWidth / 2) * W
        : layer.align === 'right' ? (0.5 + layer.maxWidth / 2) * W
          : layer.x * W;
      if (layer.align === 'left') originX = layer.x * W;
      if (layer.align === 'right') originX = layer.x * W;

      var startY = layer.y * H - ((lines.length - 1) * lineHeight) / 2;
      for (var i = 0; i < lines.length; i++) {
        ctx.fillText(lines[i], originX, startY + i * lineHeight);
      }
      ctx.shadowBlur = 0;
      ctx.shadowOffsetY = 0;
    },

    rect: function (ctx, layer, env, W, H) {
      var w = layer.w * W, h = layer.h * H;
      ctx.save();
      ctx.translate(layer.x * W, layer.y * H);
      if (layer.rotate) ctx.rotate((layer.rotate * Math.PI) / 180);
      ctx.fillStyle = layer.color;
      roundRect(ctx, -w / 2, -h / 2, w, h, layer.radius * W);
      ctx.fill();
      ctx.restore();
    },

    circle: function (ctx, layer, env, W, H) {
      ctx.fillStyle = layer.color;
      ctx.beginPath();
      ctx.arc(layer.x * W, layer.y * H, layer.r * W, 0, Math.PI * 2);
      ctx.fill();
    },

    waveform: function (ctx, layer, env, W, H, t) {
      var seconds = t / 1000;
      ctx.lineWidth = layer.thickness * W;
      ctx.strokeStyle = layer.color;
      ctx.lineCap = 'round';
      for (var band = 0; band < layer.bands; band++) {
        ctx.globalAlpha = env.alpha * (1 - band * 0.22);
        ctx.beginPath();
        var offset = band * 0.05;
        for (var px = 0; px <= W; px += 6) {
          var phase = (px / W) * layer.frequency * Math.PI * 2;
          var y = layer.y * H
            + Math.sin(phase + seconds * layer.speed * 2 + offset * 10)
            * layer.amplitude * H * (1 - band * 0.25);
          if (px === 0) ctx.moveTo(px, y); else ctx.lineTo(px, y);
        }
        ctx.stroke();
      }
      ctx.globalAlpha = env.alpha;
    },

    grid: function (ctx, layer, env, W, H, t) {
      var seconds = t / 1000;
      var cell = layer.cell * W;
      ctx.strokeStyle = layer.color;
      ctx.lineWidth = Math.max(1, W * 0.0015);
      ctx.beginPath();

      if (layer.perspective) {
        var horizon = layer.horizon * H;
        var scroll = (seconds * layer.speed * cell) % cell;
        for (var x = -W; x <= W * 2; x += cell) {
          ctx.moveTo(W / 2 + (x - W / 2) * 0.15, horizon);
          ctx.lineTo(x, H);
        }
        for (var i = 0; i < 26; i++) {
          var depth = i / 26;
          var y = horizon + Math.pow(depth, 2.4) * (H - horizon) + scroll * depth;
          if (y > horizon && y < H) { ctx.moveTo(0, y); ctx.lineTo(W, y); }
        }
      } else {
        var shift = (seconds * layer.speed * cell) % cell;
        for (var gx = -cell + shift; gx < W + cell; gx += cell) {
          ctx.moveTo(gx, 0); ctx.lineTo(gx, H);
        }
        for (var gy = -cell + shift; gy < H + cell; gy += cell) {
          ctx.moveTo(0, gy); ctx.lineTo(W, gy);
        }
      }
      ctx.stroke();
    },

    particles: function (ctx, layer, env, W, H, t) {
      var rand = mulberry32(layer.seed || 7);
      var seconds = t / 1000;
      var radius = layer.size * W;
      ctx.fillStyle = layer.color;
      for (var i = 0; i < layer.count; i++) {
        var bx = rand();
        var by = rand();
        var speed = 0.4 + rand() * 1.2;
        var y = (by - seconds * layer.speed * speed * 0.08) % 1;
        if (y < 0) y += 1;
        var wobble = Math.sin(seconds * speed + i) * W * 0.01;
        ctx.globalAlpha = env.alpha * (0.25 + rand() * 0.75);
        ctx.beginPath();
        ctx.arc(bx * W + wobble, y * H, radius, 0, Math.PI * 2);
        ctx.fill();
      }
      ctx.globalAlpha = env.alpha;
    },

    scanlines: function (ctx, layer, env, W, H) {
      var gap = Math.max(2, layer.gap * H);
      ctx.fillStyle = layer.color;
      ctx.globalAlpha = env.alpha * layer.strength;
      for (var y = 0; y < H; y += gap * 2) {
        ctx.fillRect(0, y, W, gap);
      }
      ctx.globalAlpha = env.alpha;
    },

    progress: function (ctx, layer, env, W, H, t, duration) {
      var h = layer.thickness * H;
      ctx.fillStyle = layer.color;
      ctx.globalAlpha = env.alpha * 0.25;
      ctx.fillRect(0, layer.y * H, W, h);
      ctx.globalAlpha = env.alpha;
      ctx.fillRect(0, layer.y * H, W * clamp(t / duration, 0, 1), h);
    }
  };

  function ScenePlayer(canvas, spec) {
    this.canvas = canvas;
    this.spec = spec;
    this.ctx = canvas.getContext('2d');
    this.raf = null;
    this.startedAt = 0;
    this.pausedAt = 0;
    this.width = 0;
    this.height = 0;
  }

  ScenePlayer.prototype.resize = function () {
    var dpr = Math.min(global.devicePixelRatio || 1, 2);
    var rect = this.canvas.getBoundingClientRect();
    var w = Math.max(1, Math.round(rect.width * dpr));
    var h = Math.max(1, Math.round(rect.height * dpr));
    if (w !== this.canvas.width || h !== this.canvas.height) {
      this.canvas.width = w;
      this.canvas.height = h;
    }
    this.width = w;
    this.height = h;
  };

  ScenePlayer.prototype.drawAt = function (t) {
    this.resize();
    var ctx = this.ctx, W = this.width, H = this.height;
    var duration = this.spec.duration_ms;

    paintBackground(ctx, this.spec.bg, W, H);

    for (var i = 0; i < this.spec.layers.length; i++) {
      var layer = this.spec.layers[i];
      var env = envelope(layer, t, W, H);
      if (!env || env.alpha <= 0.001) continue;
      var painter = painters[layer.type];
      if (!painter) continue;

      ctx.save();
      ctx.globalAlpha = env.alpha;
      if (env.dx || env.dy) ctx.translate(env.dx, env.dy);
      if (env.scale !== 1) {
        var ax = (layer.x == null ? 0.5 : layer.x) * W;
        var ay = (layer.y == null ? 0.5 : layer.y) * H;
        ctx.translate(ax, ay);
        ctx.scale(env.scale, env.scale);
        ctx.translate(-ax, -ay);
      }
      try {
        painter(ctx, layer, env, W, H, t, duration);
      } catch (err) {
        /* One bad layer must not blank the whole clip. */
      }
      ctx.restore();
    }
  };

  ScenePlayer.prototype.frame = function (now) {
    if (!this.startedAt) this.startedAt = now;
    var elapsed = now - this.startedAt;
    var duration = this.spec.duration_ms;
    var t = this.spec.loop ? elapsed % duration : Math.min(elapsed, duration);
    this.drawAt(t);
    this.raf = global.requestAnimationFrame(this.frame.bind(this));
  };

  ScenePlayer.prototype.play = function () {
    if (this.raf) return;
    this.startedAt = 0;
    this.raf = global.requestAnimationFrame(this.frame.bind(this));
  };

  ScenePlayer.prototype.pause = function () {
    if (this.raf) {
      global.cancelAnimationFrame(this.raf);
      this.raf = null;
    }
  };

  ScenePlayer.prototype.destroy = function () {
    this.pause();
    this.spec = null;
  };

  /* A single still frame, used for thumbnails on profile grids. */
  ScenePlayer.prototype.poster = function (atMs) {
    this.drawAt(atMs == null ? this.spec.duration_ms * 0.45 : atMs);
  };

  global.ScenePlayer = ScenePlayer;
})(window);
